libcaf_core/caf/flow/op/retry.hpp

```diff
@@ -0,0 +1,171 @@
+// This file is part of CAF, the C++ Actor Framework. See the file LICENSE in
+// the main distribution directory for license terms and copyright or visit
+// https://github.com/actor-framework/actor-framework/blob/master/LICENSE.
+
+#pragma once
+
+#include "caf/config.hpp"
+#include "caf/detail/assert.hpp"
+#include "caf/flow/backpressure_overflow_strategy.hpp"
+#include "caf/flow/observer.hpp"
+#include "caf/flow/op/hot.hpp"
+#include "caf/flow/subscription.hpp"
+
+#include <deque>
+#include <utility>
+
+namespace caf::flow::op {
+
+template <class T, class Predicate>
+class retry_sub : public subscription::impl_base, public observer_impl<T> {
+public:
+  // -- constructors, destructors, and assignment operators --------------------
+
+  retry_sub(coordinator* parent, observer<T> out, Predicate predicate,
+            observable<T> in)
+    : parent_(parent),
+      out_(std::move(out)),
+      in_(std::move(in)),
+      predicate(std::move(predicate)) {
+    // nop
+  }
+
+  // -- implementation of subscription -----------------------------------------
+
+  coordinator* parent() const noexcept override {
+    return parent_;
+  }
+
+  bool disposed() const noexcept override {
+    return !out_;
+  }
+
+  void request(size_t new_demand) override {
+    if (new_demand == 0)
+      return;
+    demand_ += new_demand;
+  }
+
+  // -- implementation of observer_impl ----------------------------------------
+
+  void ref_coordinated() const noexcept override {
+    ref();
+  }
+
+  void deref_coordinated() const noexcept override {
+    deref();
+  }
+
+  void on_subscribe(subscription sub) override {
+    if (sub_) {
+      sub_.cancel();
+      sub_.release_later();
+    }
+    sub_ = std::move(sub);
+    sub_.request(demand_);
+  }
+
+  void on_next(const T& item) override {
+    if (!out_)
+      return;
+    if (demand_ > 0) {
+      --demand_;
+      out_.on_next(item);
+      if (sub_)
+        sub_.request(1);
+      return;
+    }
+  }
```

Comment:
The operator must not proactively hand out demand. The observer must drive demand, not this operator.

Calling `sub_.request` should (only) be done in two cases:
- `request` is called and `sub_` is valid
- `on_subscribe` is called and `demand_` is > 0